### PR TITLE
Microbe damage effects

### PR DESCRIPTION
**Pull request by crodnu**
Adding a damage effect to the membranes

### CODE REVIEW

**Comment by hhyyrylainen** [comment]
There is still one formatting issue:
```diff
--- a/scripts/microbe_stage/microbe.as
+++ b/scripts/microbe_stage/microbe.as
@@ -504,7 +504,7 @@ class MicrobeSystem : ScriptSystem{
         applyCellMovement(components, logicTime);
 
         compoundAbsorberComponent.setAbsorbtionCapacity(microbeComponent.capacity);
-        
+
         membraneComponent.setHealthPercentage(microbeComponent.hitpoints / microbeComponent.maxHitpoints);
     }
```

File: scripts/microbe_stage/microbe.as [hunk 1]
```diff
@@ -504,6 +504,8 @@ class MicrobeSystem : ScriptSystem{
         applyCellMovement(components, logicTime);
 
         compoundAbsorberComponent.setAbsorbtionCapacity(microbeComponent.capacity);
+        
+        membraneComponent.setHealthPercentage(microbeComponent.hitpoints / microbeComponent.maxHitpoints);
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe.as]: There should be a check here to see if the value changed in order to not update the material of *all* cells on each frame, which I think this version does.

File: src/microbe_stage/membrane_system.cpp [hunk 2]
```diff
@@ -206,6 +207,25 @@ void
     }
 }
 
+void
+    MembraneComponent::setHealthPercentage(float value)
+{
+    healthPercentage = std::clamp<float>(value, 0.0, 1.0);
```
Comment (hhyyrylainen) [inline src/microbe_stage/membrane_system.cpp]: Would this line work like this: `std::clamp(value, 0.0f, 1.0f);` (all floats, so template type deduction should work)

File: src/microbe_stage/membrane_system.cpp [hunk 3]
```diff
@@ -206,6 +207,25 @@ void
...
+    // If we already have created a material we need to re-apply it
+    if(coloredMaterial) {
+        coloredMaterial->getTechnique(0)
+            ->getPass(0)
+            ->getFragmentProgramParameters()
+            ->setNamedConstant("healthPercentage", healthPercentage);
+        coloredMaterial->getTechnique(0)
+            ->getPass(0)
+            ->getTextureUnitState(0)
+            ->setHardwareGammaEnabled(true);
```
Comment (hhyyrylainen) [inline src/microbe_stage/membrane_system.cpp]: Why is this changed here?

File: src/microbe_stage/membrane_system.h [hunk 4]
```diff
@@ -53,6 +53,10 @@ class MembraneComponent : public Leviathan::Component {
     void
         setColour(const Float4& value);
 
+    //! Should set the health percentage.
+    void
+        setHealthPercentage(float value);
```
Comment (hhyyrylainen) [inline src/microbe_stage/membrane_system.h]: The health value is between 0 and 1. Percentages go up to 100. This should be renamed to like health fraction and a comment should be added about the valid range.

File: src/microbe_stage/membrane_system.cpp (hunk 3 above)
Comment (crodnu) [reply src/microbe_stage/membrane_system.cpp]: wdym

Comment (crodnu) [reply src/microbe_stage/membrane_system.cpp]: that was just copy-pasted from the colour change method above, it might not be needed.

Comment (hhyyrylainen) [reply src/microbe_stage/membrane_system.cpp]: Why this is here:
```
setHardwareGammaEnabled(true);
```
Why it is needed? If this is needed somewhere it very much should be in the material creation.

Comment (crodnu) [reply src/microbe_stage/membrane_system.cpp]: ¯\_(ツ)_/¯
maybe it should be removed from the color change too?

Comment (hhyyrylainen) [reply src/microbe_stage/membrane_system.cpp]: I think it should. I'm working on other things ATM, but if that isn't in the material initialization place it should be moved there and removed from everywhere else.

**Comment by crodnu** [comment]
Oh you already approved of the changes? I guess the browser cached an old page or something...

**Comment by Untrustedlife** [comment]
SO is this okay to merge now?

**Comment by hhyyrylainen** [comment]
I fixed the non-squashed merge, I hope.
